script/release

```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+# Usage: script/release [<newversion> | major | minor | patch | premajor | preminor | prepatch | prerelease | from-git]
+#
+# - Triggers a version bump to the provided explicit version,
+#   or to the derived new version given the "bump level".
+#   Leverages npm-version to the bumping and git-commit/tag.
+#   Requires git to be configured for GPG signing.
+
+set -euo pipefail
+IFS=$'\n\t'
+
+BUMP=${1:-patch}
+RELNOTES="$(git rev-parse --git-dir)/RELNOTES"
+
+./script/release-notes
+
+printf -v MESSAGE 'Bats %%s\n\n%s' "$(sed -E 's/^#+[ \t]*//' "$RELNOTES")"
```

Comment:
Maybe replace `[  \t]` with `[[:space:]]` just for readability's sake. That might render the `-E` unnecessary in the process.

Reply:
👍 I think the `-E` is required for the `+` still, though we could get around that with `##*`, right?

Reply:
Oh, is that what needs it? Then feel free to leave it, then.

Reply:
You can actually get around the need for `-E` by escaping the `+` with a backslash.  It works similarly to `grep` in that regard.

Reply:
Won't backslash-escaping just match the literal `+` character, though, instead of interpreting it as the symbol for "one or more of the previous expression"?

Reply:
Not with `sed` if you're not using `-E`.  `sed` actually requires backslash escaping for `+` and parentheses (at minimum - might be others, too) if `-E` is not used.

From a quick demo to show it on an old Ubuntu 14.04 box I have up:

```
root@ubuntu:~# cat /etc/issue
Ubuntu 14.04.4+ LTS \n \l
Testing parens ()
root@ubuntu:~# sed 's/[0-9]+/replaced/g' /etc/issue
Ubuntu 14.04.replaced LTS \n \l
Testing parens ()
root@ubuntu:~# sed 's/[0-9]\+/replaced/g' /etc/issue
Ubuntu replaced.replaced.replaced+ LTS \n \l
Testing parens ()
root@ubuntu:~# sed 's/\(1\)/words/g' /etc/issue
Ubuntu words4.04.4+ LTS \n \l
Testing parens ()
root@ubuntu:~# sed 's/()/words/g' /etc/issue
Ubuntu 14.04.4+ LTS \n \l
Testing parens words
root@ubuntu:~# sed -E 's/[0-9]+/replaced/g' /etc/issue
Ubuntu replaced.replaced.replaced+ LTS \n \l
Testing parens ()
root@ubuntu:~# sed -E 's/(1)/words/g' /etc/issue
Ubuntu words4.04.4+ LTS \n \l
Testing parens ()
root@ubuntu:~#
```

Reply:
Wow. Very good to know. Just when I thought I knew regexes, and `sed`...

Reply:
I'm simultaneously happy to TIL, and sad that such a "basic" thing still has so much TIL-able left...

Speaking more generally though, I see the appeal to avoid `-E` because of broad appeals to minimalism. But at the same time, I feel like using `-E` _always_ reduces the risk for future gotchas. (I'm assuming that most everyone has a more solid understanding of the extended form, and are less likely to introduce bugs if all regexes are already forcing extended mode.) Are there any benefits to avoiding `-E`?

Reply:
I don't mind keeping it, now that I'm surprised to learn that certain basic regex constructs actually _require_ it to work as I'd expected.

Reply:
I don't think it's a problem.  As far as I am aware `-E` (for both `sed` and `grep`) just changes which regex engine is used.  I don't think that opens us up to anything particularly nasty like `echo`'s `-e` does.

Reply:
I'd also like to note that it was implemented in this manner for the same reason many things are done the way they are with core tools on *nix systems: to maintain backwards compatibility with scripts.